### PR TITLE
feat(SPV-1391): add source transaction builder. 

### PR DESCRIPTION
**Pull request by mgosek-4chain**
[SPV-1391-graph-tx-model.pdf](https://github.com/user-attachments/files/18571803/SPV-1391-graph-tx-model.pdf)

Unit test cases implemented  ✅: 
- 1, 2, 3, 4, 5, 6, 7

# Pull Request Checklist

- [X] 📖 I created my PR using provided  : [CODE_STANDARDS](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_STANDARDS.md)
- [X] 📖 I have read the short Code of Conduct: [CODE_OF_CONDUCT](https://github.com/bitcoin-sv/spv-wallet/blob/main/.github/CODE_OF_CONDUCT.md)
- [X] 🏠 I tested my changes locally.
- [X] ✅ I have provided tests for my changes.
- [X] 📝 I have used conventional commits.
- [X] 📗 I have updated any related documentation.
- [X] 💾 PR was issued based on the Github or Jira issue.

### CODE REVIEW

File: engine/transaction/outlines/internal/inputs/source_transaction_builder_test.go [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+			RawHex:     Ptr(tx1Hex),
+		},
+		{
+			SourceTXID: tx1InputIdx0SourceID,
+			BeefHex:    Ptr(tx6BEEFHex),
+		},
+	})
+
+	// then:
+	require.NoError(t, err)
```
Comment (chris-4chain) [inline engine/transaction/outlines/internal/inputs/source_transaction_builder_test.go:47]: I think that more complex assertions regarding generated `SourceTransactionBuilder::Tx` is crucial.

File: engine/transaction/outlines/internal/inputs/source_transaction_builder.go [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+			return nil, fmt.Errorf("failed to add entry to source transaction map: %w", err)
+		}
+	}
+	return sourceTXs, nil
+}
+
+// SourceTransactionBuilder builds transactions from TxQueryResults.
+type SourceTransactionBuilder struct {
+	Tx *sdk.Transaction // Root transaction.
+}
```
Comment (chris-4chain) [inline engine/transaction/outlines/internal/inputs/source_transaction_builder.go:114]: What is the purpose of this builder?
- When will it be executed? (While receiving or sending transaction)
- What is the outcome?, (I mean the `Tx` field). I guess that it is a fully sourced transaction up to `mined` ancestors.

File: engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+const (
+	DefaultBlockHeight = 1000
+	DefaultSatoshis    = 100
+)
+
+// TransactionDescriptor contains a transaction and its associated hexadecimal data.
+type TransactionDescriptor struct {
+	Tx      *sdk.Transaction // The actual transaction.
+	BEEFHex *string          // Hexadecimal representation of the transaction in BEEF format.
+	RawHex  *string          // Optional raw hexadecimal representation of the transaction.
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go:28]: Hmmm. This implementation has one potential issue.
Because of the fields are exported, someone is able to set both `BEEF` and `Raw` strings. Then `IsBEEF` and `IsRaw` will return true.

I don't know if this is a problem 🤷 - maybe not. 
But, if so, I would recommend to set `format` enum.

```go
type Format int

const (
	Raw Format = iota
	BEEF
)

type TransactionDescriptor struct {
	Tx     *sdk.Transaction // The actual transaction.
	Format Format
	Hex    string
}
```

Comment (chris-4chain) [reply engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go:28]: ...or :D 
Maybe you could make `TransactionDescriptor` unexported and then it would be enough.

File: engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+// IsBeef checks if the actual transaction was serialized to HexBEEF format.
+func (t TransactionDescriptor) IsBEEF() bool { return t.BEEFHex != nil }
+
+// IsRaw checks if the actual transaction was serialized to Hex format.
+func (t TransactionDescriptor) IsRaw() bool { return t.RawHex != nil }
+
+// GraphBuilderTransactions is a map where the key is a transaction name (string)
+// and the value is a TransactionDescriptor containing details of the transaction.
+type GraphBuilderTransactions map[string]*TransactionDescriptor
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go:44]: Why pointer to `TransactionDescriptor`. 
I'm asking because I noticed that you defined all its method as "value - receivers".

File: engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+// AddRawTx adds a raw transaction to the GraphBuilderTransactions map.
+// The name is used as the key, and the transaction is stored with its hexadecimal representation.
+func (g GraphBuilderTransactions) AddRawTx(t *testing.T, name string, tx *sdk.Transaction) {
+	t.Helper()
+	require.NotEmpty(t, name, "Name of the raw transaction should not be empty")
+	require.NotNil(t, tx, "Added transaction should not be nil")
+
+	hex := tx.Hex()
+	g[name] = &TransactionDescriptor{Tx: tx, RawHex: &hex}
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go:54]: Maybe some checking if this name already exist in map. 
That would help during development of new test cases.

File: engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+	transactions         GraphBuilderTransactions // Collection of transactions managed by the builder.
+	p2pKHLockingScript   *script.Script           // Locking script for P2PKH transactions.
+	p2pKHUnlockingScript *p2pkh.P2PKH             // Unlocking script for P2PKH transactions.
+	hexGen               HexGen                   // Generator for unique hexadecimal values.
+	blockHeight          BlockHeight              // Current block height.
+	satoshis             uint64                   // Default amount of satoshis for mined transactions.
+}
+
+// GraphBuilderTransactions returns the map of transactions managed by the builder.
+func (g *GraphBuilder) GraphBuilderTransactions() GraphBuilderTransactions {
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go:83]:
```suggestion
func (g *GraphBuilder) Transactions() GraphBuilderTransactions {
```

because the call looks redundant, e.g.: `graphBuilder.GraphBuilderTransactions()`

File: engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+// HexGen generates unique hexadecimal values for transactions.
+type HexGen int32
+
+// Inc increments the hexadecimal generator to produce the next value.
+func (h *HexGen) Inc() { *h++ }
+
+// Val returns the current hexadecimal value as a string.
+func (h HexGen) Val() string {
+	return fmt.Sprintf("a%010de1b81dd2c9c0c6cd67f9bdf832e9c2bb12a1d57f30cb6ebbe78d9", h)
+}
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go:170]: 🤷 
image

Comment (mgosek-4chain) [reply engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go]: Fixed :)

Comment (chris-4chain) [reply engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go]: Hmmm I don't see, it's fixed. Maybe you forgot to push your local changes?

File: engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go (hunk 3 above)
Comment (chris-4chain) [reply engine/v2/transaction/outlines/testabilities/txgraph/graph_builder.go]: What we're doing with this?

File: engine/v2/transaction/outlines/source_transaction_builder.go [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+// Add adds a TxQueryResult to the SourceTxMap after parsing it.
+func (m SourceTxMap) Add(q *TxQueryResult) error {
+	if q == nil || q.IsZero() {
+		return nil
+	}
+
+	if q.IsBeef() && q.IsRawTx() {
+		return ErrMutuallyExclusiveTxQueryResult
+	}
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/source_transaction_builder.go:71]: If you used the approach I mention earlier ☝️ , you wouldn't have such problem.

File: engine/v2/transaction/outlines/source_transaction_builder.go [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+// Value retrieves the SourceTx for a given ID or returns an empty SourceTx.
+func (m SourceTxMap) Value(id string) SourceTx {
+	return m[id]
+}
+
+// Add adds a TxQueryResult to the SourceTxMap after parsing it.
+func (m SourceTxMap) Add(q *TxQueryResult) error {
+	if q == nil || q.IsZero() {
+		return nil
+	}
```
Comment (chris-4chain) [inline engine/v2/transaction/outlines/source_transaction_builder.go:67]: Hmmm, I think you mixing "pointer based" approach and "value based".
Methods are with "value receivers" but here the method `Add` gets pointer to it. And then you check `isNil` and `IsZero`.
Maybe, stick to one approach. 

... Or, you did it on purpose (e.g. for performance reasons) and I don't see it 🙈 😅

**Review by dorzepowski — CHANGES_REQUESTED** [review]

**Review by dorzepowski — DISMISSED** [review]

File: engine/v2/transaction/outlines/source_transaction_builder.go [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+}
+
+// Add adds a TxQueryResult to the SourceTxMap after parsing it.
+func (m SourceTxMap) Add(q *TxQueryResult) error {
+	if q == nil {
+		return nil
+	}
+
+	if q.IsBeef() && q.IsRawTx() {
+		return ErrMutuallyExclusiveTxQueryResult
```
[the remark this answers is not in the document]
Comment (mgosek-4chain) [reply engine/v2/transaction/outlines/source_transaction_builder.go:65]: This has been added to detect cases when mapping in the repository layer was done inappropriate.

File: engine/v2/transaction/outlines/source_transaction_builder_test.go [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+
+	// subject tx:
+	subjectTx := sdk.NewTransaction()
+	subjectTx.AddInput(&sdk.TransactionInput{SourceTXID: ID})
+
+	db := toTxQueryResultSlice(txgraph.GraphBuilderTransactions{})
+
+	// when:
+	builder := SourceTransactionBuilder{Tx: subjectTx}
+	err = builder.Build(db)
```
[the remark this answers is not in the document]
Comment (mgosek-4chain) [reply engine/v2/transaction/outlines/source_transaction_builder_test.go:27]: My goal was to keep the PR minimal by introducing a subcomponent that will eventually be used as part of the mentioned service. However, I can create the full service with placeholders if that would be preferable. Let’s discuss which approach makes the most sense 🤔

File: engine/v2/transaction/outlines/source_transaction_builder.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+	if b.Tx == nil {
+		return ErrNilTxBuilder
+	}
+
+	sourceTxs, err := res.SourceTxMap()
+	if err != nil {
+		return fmt.Errorf("failed to convert query results to source transaction map: %w", err)
+	}
+
+	b.buildRecursive(b.Tx.Inputs, sourceTxs)
```
[the remark this answers is not in the document]
Comment (mgosek-4chain) [reply engine/v2/transaction/outlines/source_transaction_builder.go:119]: Definitely! This will make the method call much simpler.

File: engine/v2/transaction/outlines/source_transaction_builder_test.go [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+	// then:
+	require.NoError(t, err)
+}
+
+func TestSourceTransactionBuilder_CommonBEEFTxGreatGrandparentsForTx1Tx3(t *testing.T) {
+	// given:
+	graphBuilder := txgraph.NewGraphBuilder(t)
+
+	// graph:
+	tx6 := graphBuilder.CreateMinedTx("tx6", 2)
```
[the remark this answers is not in the document]
Comment (mgosek-4chain) [reply engine/v2/transaction/outlines/source_transaction_builder_test.go:208]: To avoid overriding the current values stored in the graph.

File: engine/v2/transaction/beef/beef_service.go [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+// IsRawTx checks if the TxQueryResult contains a raw transaction.
+func (tx TxQueryResult) IsRawTx() bool { return tx.RawHex != nil }
+
+// TxRepository defines an interface for querying transaction input sources.
+// It provides a method to retrieve transaction details for a given set of source transaction IDs.
+type TxRepository interface {
+	// QueryInputSources retrieves transaction query results for the provided source transaction IDs.
+	// Returns a slice of TxQueryResult and an error if the query fails.
+	QueryInputSources(ctx context.Context, sourceTXIDs ...string) (TxQueryResultSlice, error)
+}
```
Comment (chris-4chain) [inline engine/v2/transaction/beef/beef_service.go:30]: I recall we established that such interfaces should be in `interface.go` file in at the particular service level.

File: engine/v2/transaction/beef/beef_service.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+func (s *Service) extractSourceTXIDs(tx *sdk.Transaction) ([]string, error) {
+	if tx.InputCount() == 0 {
+		return nil, txerrors.ErrZeroInputCount
+	}
+
+	sourceTXIDs := make([]string, 0)
+	for _, in := range tx.Inputs {
+		sourceTXIDs = append(sourceTXIDs, in.SourceTXID.String())
+	}
+	return sourceTXIDs, nil
```
Comment (chris-4chain) [inline engine/v2/transaction/beef/beef_service.go:48]: Why not reserve memory in the slice if you know the length?



Personally, I would use here iterators to not waste memory for intermediate slice - ... but I'm weirdo 🙈

File: engine/v2/transaction/beef/beef_service_test.go [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+	// given:
+	graphBuilder := testabilities.NewTxGraphBuilder(t)
+	repository := testabilities.NewTxRepository(t, graphBuilder)
+	service := beef.NewService(repository)
+
+	var hexGen testabilities.HexGen
+	ID, err := chainhash.NewHashFromHex(hexGen.Val())
+	require.NoError(t, err, "expected to create hash from the hex gen val: %s", hexGen.Val())
+
+	// subject tx:
```
Comment (chris-4chain) [inline engine/v2/transaction/beef/beef_service_test.go:25]: **Very minor**
Hmmm, this comment breaks `given, when, then` pattern but doesn't add additional context because one line below there is `subjectTx` variable. I would skip the comment.

**Review by chris-4chain — APPROVED** [review]

File: engine/v2/transaction/beef/beef_service.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+
+	txID := tx.TxID().String()
+
+	// Extract source transaction IDs from the provided transaction.
+	sourceTxIDs, err := s.extractSourceTXIDs(tx)
+	if err != nil {
+		return "", fmt.Errorf("failed to extract source transaction IDs for transaction %s: %w", txID, err)
+	}
+
+	// Query the repository for source transactions.
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/beef_service.go:67]: remove redundant comments please

File: engine/v2/transaction/beef/beef_service_test.go [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+
+	// when:
+	hexBEEF, err := service.PrepareBEEF(context.Background(), subjectTx)
+
+	// then:
+	require.NoError(t, err)
+	require.NotEmpty(t, hexBEEF)
+
+	testabilities.AssertTxInputs(t, subjectTx.Inputs...)
+	testabilities.AssertTxBEEFHex(t, subjectTx)
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/beef_service_test.go:61]: It would be good to follow the pattern of testabilities from other packages.
But it could be done in different PR.

File: engine/v2/transaction/beef/testabilities/tx_scripts.go [hunk 19]
```diff
@@ -0,0 +1,73 @@
...
+	bip32 "github.com/bitcoin-sv/go-sdk/compat/bip32"
+	primitives "github.com/bitcoin-sv/go-sdk/primitives/ec"
+	"github.com/bitcoin-sv/go-sdk/script"
+	"github.com/bitcoin-sv/go-sdk/transaction/template/p2pkh"
+	"github.com/stretchr/testify/require"
+)
+
+// TxScripts is a utility structure that provides access to transaction-related scripts
+// and keys for creating and validating Bitcoin transactions.
+type TxScripts struct {
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/testabilities/tx_scripts.go:15]: If it's not used outsie this package, I would make it private

File: engine/v2/transaction/beef/beef_service_test.go [hunk 20]
```diff
@@ -0,0 +1,258 @@
...
+	subjectTx.AddInput(&sdk.TransactionInput{SourceTXID: tx0.Inputs[0].SourceTXID})
+
+	// when:
+	hexBEEF, err := service.PrepareBEEF(context.Background(), subjectTx)
+
+	// then:
+	require.NoError(t, err)
+	require.NotEmpty(t, hexBEEF)
+
+	testabilities.AssertTxInputs(t, subjectTx.Inputs...)
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/beef_service_test.go:60]: Those assertions should be rather done on transaction parsed from received beef

File: engine/v2/transaction/beef/beef_service.go [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+	if tx == nil {
+		return "", txerrors.ErrNilSubjectTx
+	}
+
+	txID := tx.TxID().String()
+
+	// Extract source transaction IDs from the provided transaction.
+	sourceTxIDs, err := s.extractSourceTXIDs(tx)
+	if err != nil {
+		return "", fmt.Errorf("failed to extract source transaction IDs for transaction %s: %w", txID, err)
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/beef_service.go:64]: Please use `spverrors.Wrapf`,
maybe it's not the best solution but at least it is consistent with the rest of repository, and this could make it easier for us to replace it later.

File: engine/v2/transaction/beef/source_transaction_resolver.go [hunk 22]
```diff
@@ -0,0 +1,145 @@
...
+// Add adds a TxQueryResult to the SourceTxMap after parsing it.
+func (m SourceTxMap) Add(q *TxQueryResult) error {
+	if q == nil {
+		return txerrors.ErrNilTxQueryResult
+	}
+
+	if q.IsBeef() {
+		tx, err := sdk.NewTransactionFromBEEFHex(*q.BeefHex)
+		if err != nil {
+			return fmt.Errorf("failed to parse BEEF transaction: %w", err)
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/source_transaction_resolver.go:45]: Shouldn't that be a public error (from errors.go file)?

File: engine/v2/transaction/beef/source_transaction_resolver.go [hunk 23]
```diff
@@ -0,0 +1,145 @@
...
+		}
+		m[q.SourceTXID] = SourceTx{Tx: tx, HadBeef: true}
+		return nil
+	}
+
+	if q.IsRawTx() {
+		tx, err := sdk.NewTransactionFromHex(*q.RawHex)
+		if err != nil {
+			return fmt.Errorf("failed to parse raw transaction: %w", err)
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/source_transaction_resolver.go:54]: Shouldn't that be a public error (from errors.go file)?

File: engine/v2/transaction/beef/source_transaction_resolver.go [hunk 24]
```diff
@@ -0,0 +1,145 @@
...
+
+// TxQueryResultSlice represents a slice of TxQueryResult.
+type TxQueryResultSlice []*TxQueryResult
+
+// SourceTxMap converts the slice to a SourceTxMap.
+func (qq TxQueryResultSlice) SourceTxMap() (SourceTxMap, error) {
+	sourceTXs := make(SourceTxMap)
+	for _, q := range qq {
+		if err := sourceTXs.Add(q); err != nil {
+			return nil, fmt.Errorf("failed to add entry to source transaction map: %w", err)
```
Comment (dorzepowski) [inline engine/v2/transaction/beef/source_transaction_resolver.go:76]: (Again) please use in this file `spverrors.Wrapf` instead, to keep consistency in code.

File: engine/v2/transaction/beef/beef_service.go [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// extractSourceTXIDs extracts source transaction IDs from the inputs of the given transaction.
+// It returns a slice of transaction IDs or an error if the transaction has no inputs.
+func (s *Service) extractSourceTXIDs(tx *sdk.Transaction) ([]string, error) {
+	if tx.InputCount() == 0 {
+		return nil, txerrors.ErrZeroInputCount
+	}
+
+	sourceTXIDs := make([]string, len(tx.Inputs))
```
Comment (wregulski) [inline engine/v2/transaction/beef/beef_service.go:44]: I am not sure but doesn't mixing prealloacting with "append" function cause memory leaks sometimes?

What I mean:
```go
foos := make([]string, len(items))
for _, item := range items {
      foos = append(foos, "Foo" + item) // Appending to pre-allocated slice
}
```

to be replaced with:
```go
foos := make([]string, len(items))
for i, item := range items {
     foos[i] = "Foo" + item // Direct assignment using index
}
```

File: engine/v2/transaction/beef/beef_service_test.go [hunk 26]
```diff
@@ -0,0 +1,272 @@
...
+
+	"github.com/bitcoin-sv/go-sdk/chainhash"
+	sdk "github.com/bitcoin-sv/go-sdk/transaction"
+	"github.com/bitcoin-sv/spv-wallet/engine/v2/transaction/beef"
+	"github.com/bitcoin-sv/spv-wallet/engine/v2/transaction/beef/testabilities"
+	txerrors "github.com/bitcoin-sv/spv-wallet/engine/v2/transaction/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func TestPrepareBEEF_MissingSourceTx_InSubjectTxInput(t *testing.T) {
```
Comment (wregulski) [inline engine/v2/transaction/beef/beef_service_test.go:15]: Maybe let's add more edge cases to those tests (I would like to cover all paths in tests).
For example:
-> when source ID from input is invalid hex,
-> when BEEF has invalid hex format,
-> malformed transaction structure in PrepareBEEF

**Review by wregulski — CHANGES_REQUESTED** [review]

File: engine/v2/transaction/beef/testabilities/tx_repository.go [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+	"github.com/bitcoin-sv/spv-wallet/engine/v2/transaction/beef"
+	"github.com/stretchr/testify/require"
+)
+
+type TxRepository struct {
+	graph *TxGraphBuilder
+}
+
+func (t *TxRepository) QueryInputSources(ctx context.Context, sourceTXIDs ...string) (beef.TxQueryResultSlice, error) {
+	return t.graph.ToTxQueryResultSlice(), nil
```
[the remark this answers is not in the document]
Comment (chris-4chain) [reply engine/v2/transaction/beef/testabilities/tx_repository.go:16]: Hmmm, it's just mocked repository implementation. 
I think, it's enough that the interface defines these methods with `ctx` (which will be utilized in actual repository implementation).

[the remark this answers is not in the document]
Comment (wregulski) [reply engine/v2/transaction/beef/testabilities/tx_repository.go:16]: You are right

**Review by wregulski — APPROVED** [review]

**Review by chris-4chain — APPROVED** [review]
